### PR TITLE
Mega PR to kickstart the 4.0.0-RC1 development cycle

### PR DESCRIPTION
**Pull request by xeno-by**
Fixes 20 out of 51 tickets that are currently scheduled for 4.0.0-RC1.

### CODE REVIEW

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala [hunk 1]
```diff
@@ -159,17 +159,9 @@ object Scalacp {
...
-      else if (name.endsWith(" ")) loop(name.substring(0, name.length - 1))
-      else if (name == "<root>") n.RootPackage
-      else if (name == "<empty>") n.EmptyPackage
-      else if (name == "<init>") n.Constructor
-      else if (name == "<refinement>") "$anon"
-      else NameTransformer.decode(name)
-    }
-    loop(sym.name)
+    val ssym = ssymbol(sym)
+    if (skind(sym) == k.PACKAGE_OBJECT) ssym.owner.desc.name
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: `skind(sym).isPackageObject` (but I don't feel strongly about it)

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala [hunk 2]
```diff
@@ -159,17 +159,9 @@ object Scalacp {
...
-      if (i > 0) loop(name.substring(i + 2))
-      else if (name.endsWith(" ")) loop(name.substring(0, name.length - 1))
-      else if (name == "<root>") n.RootPackage
-      else if (name == "<empty>") n.EmptyPackage
-      else if (name == "<init>") n.Constructor
-      else if (name == "<refinement>") "$anon"
-      else NameTransformer.decode(name)
-    }
-    loop(sym.name)
+    val ssym = ssymbol(sym)
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: ⚠️ Warning: handwavy perf comment

This seems like a heavy hammer for a small problem. I see you moved `loop` to another place, I'm not sure exactly what the consequences for this refactoring are but it seems to be doing much more work now for the same functionality with a special case only for package object 🤔

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala [hunk 3]
```diff
@@ -435,11 +434,34 @@ object Scalacp {
           false
       }
     }
+    def isScalacField: Boolean = {
+      val isField = sym.isInstanceOf[MethodSymbol] && !sym.isMethod && !sym.isParam
+      val isJavaDefined = sym.isJava
+      isField && !isJavaDefined
+    }
+    def isUselessField: Boolean = {
+      val peers = sym.parent.map(_.children.toList).getOrElse(Nil)
```
Comment (olafurpg) [inline semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: ⚠️ handwavy

This seems like a heavy hammer for a small task, and on a hot path

File: semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/SymbolOps.scala [hunk 4]
```diff
@@ -125,18 +126,31 @@ trait SymbolOps { self: SemanticdbOps =>
...
 
   implicit class XtensionGScope(decls: g.Scope) {
-    def filtered: List[g.Symbol] = {
-      decls.sorted.filter { decl =>
-        !decl.isSyntheticConstructor &&
-        !decl.isLocalChild &&
-        !decl.isSyntheticValueClassCompanion
-      }
-    }
+    def useful: List[g.Symbol] = decls.sorted.filter(_.isUseful)
```
Comment (olafurpg) [inline semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/SymbolOps.scala]: `.filter(..).sorted`

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/TargetedSuite.scala [hunk 5]
```diff
@@ -873,23 +818,12 @@ class TargetedSuite extends SemanticdbSuite() {
...
        |  Bar => _empty_.ad.Bar#
        |  y => _empty_.ad.zz.$anon#y().
        |  Int => scala.Int#
-       |_empty_.ad.zz. => private[this] val field zz: Bar { val method y: Int }
-       |  Bar => _empty_.ad.Bar#
-       |  y => _empty_.ad.zz.$anon#y().
-       |  Int => scala.Int#
        |_empty_.ad.zz.$anon#y(). => val method y: : Int
        |  Int => scala.Int#
-       |_empty_.ad.zz.$anon#y. => private[this] val field y: Int
```
Comment (olafurpg) [inline tests/jvm/src/test/scala/scala/meta/tests/semanticdb/TargetedSuite.scala]: 👋 Adieu `private[this] val field`, I will not miss you

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/TargetedSuite.scala [hunk 6]
```diff
@@ -971,20 +972,16 @@ object Compat {
     s"Stream.ConsWrapper#`#::`($name)."
   }
 
-  val ListBufferDeclsInString = "+43 decls"
+  val ListBufferDeclsInString = "+39 decls"
   lazy val ListBufferDeclsActual: String = {
-    val consWrapper = typeOf[scala.collection.mutable.ListBuffer[_]]
-    val decls = consWrapper.decls.size
-    s"+$decls decls"
+    if (scalaVersion.startsWith("2.11")) "+36 decls"
```
Comment (olafurpg) [inline tests/jvm/src/test/scala/scala/meta/tests/semanticdb/TargetedSuite.scala]: note to self, we should move 2.11/2.12 specific tests to scala-2.11/scala-2.12 specific source directories

File: semanticdb/semanticdb3/src/main/scala/scala/meta/internal/semanticdb3/Locator.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package scala.meta.internal.semanticdb3
+
+import java.nio.file._
+import java.util.jar._
+import scala.collection.JavaConverters._
+
+object Locator {
+  def apply(paths: List[Path])(fn: (Path, TextDocuments) => Unit): Unit = {
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply semanticdb/semanticdb3/src/main/scala/scala/meta/internal/semanticdb3/Locator.scala]: This is the only blocking comment, the rest of the commits LGTM 👍

[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/semanticdb3/src/main/scala/scala/meta/internal/semanticdb3/Locator.scala]: Please feel free to iterate on the API yourself once we merge. I don't have a strong opinion here, as long as there's an overload or something that takes a list of paths and a `TextDocument => Unit` - that we need internally at Twitter.

File: semanticdb/semanticdb3/guide.md [hunk 8]
```diff
@@ -541,12 +541,6 @@ Caveats:
     compatibility across different versions of the `semanticdb3` library.
     We are planning to [improve the situation](https://github.com/scalameta/scalameta/issues/1300)
     in the future.
-  * At the moment, SemanticDB-based tools are responsible for implementing discovery of
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/semanticdb3/guide.md]: Done.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/Main.scala [hunk 9]
```diff
@@ -66,7 +67,9 @@ class Main(settings: Settings, reporter: Reporter) {
     classpath.foreach { entry =>
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/Main.scala]: Metacp is going through .class files, as opposed to .semanticdb files. It also needs additional caching facilities as well as parallel processing, so I figured I wouldn't overabstract.

File: semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala [hunk 10]
```diff
@@ -410,6 +412,7 @@ object Scalacp {
           false
       }
     }
+    def isLocalChild: Boolean = sym.name == "<local child>"
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala]: I know, right? 🤣

File: tests/jvm/src/test/resources/lowlevel.expect [hunk 11]
```diff
@@ -60,32 +60,179 @@ Schema => SemanticDB v3
...
+classes.C2# => final class C2.{+5 decls}
   extends AnyVal
-classes.C#`<init>`(Int). => primary ctor <init>: (val x: Int)
-  x => classes.C#`<init>`(Int).(x)
+classes.C2#`<init>`(Int). => primary ctor <init>: (val x2: Int)
+  x2 => classes.C2#`<init>`(Int).(x2)
+  Int => scala.Int#
+classes.C2#`<init>`(Int).(x2) => val param x2: Int
+  Int => scala.Int#
+classes.C2#equals(Any). => method equals: (x$1: Any): Boolean
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply tests/jvm/src/test/resources/lowlevel.expect]: 👍

File: tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala [hunk 12]
```diff
@@ -49,10 +49,6 @@ class ExpectSuite extends FunSuite with DiffAssertions {
         import MetacMetacpExpectDiffExpect._
         assertNoDiff(loadObtained, loadExpected)
       }
-      test("metac-metacp.index.diff") {
```
[the remark this answers is not in the document]
Comment (xeno-by) [reply tests/jvm/src/test/scala/scala/meta/tests/semanticdb/ExpectSuite.scala]: Very much so! That was a great idea.

**Pull request merged**